src/ir/auth_logic/ast_construction_test.cc

```diff
@@ -26,51 +26,78 @@
+          [](Attribute attribute) {
+            return absl::StrCat(attribute.principal().name(), " ",
+                                ToString(attribute.predicate()));
+          },
+          [](CanActAs can_act_as) {
+            return absl::StrCat(can_act_as.left_principal().name(),
+                                " canActAs ",
+                                can_act_as.right_principal().name());
+          }},
+      basefact.GetValue());
+}
+
+std::string ToString(Fact fact) {
+  std::string cansay_string = "";
+  for (const auto& cansay : fact.can_say()) {
+    cansay_string =
+        absl::StrCat(cansay.principal().name(), " canSay ", cansay_string);
+  }
+  return absl::StrCat(cansay_string, " ", ToString(fact.base_fact()));
+}
+
 class AstConstructionTest : public testing::TestWithParam<QueryTestData> {};
 
 TEST_P(AstConstructionTest, SimpleTestWithQueryProgram) {
-  const auto& [input_text, query_name, principal_name, fact_name, predicate] =
-      GetParam();
+  const auto& [input_text, query_name, principal, fact] = GetParam();
   Program prog_out = ParseProgram(input_text);
   ASSERT_EQ(prog_out.queries().size(), 1);
   const Query& query = prog_out.queries().front();
   EXPECT_EQ(query.name(), query_name);
-  EXPECT_EQ(query.principal().name(), principal_name);
-  EXPECT_EQ(std::get<datalog::Predicate>(
-                std::get<BaseFact>(query.fact().GetValue()).GetValue())
-                .name(),
-            fact_name);
-  EXPECT_EQ(std::get<datalog::Predicate>(
-                std::get<BaseFact>(query.fact().GetValue()).GetValue()),
-            predicate);
+  EXPECT_EQ(query.principal().name(), principal.name());
+  EXPECT_EQ(ToString(query.fact()), ToString(fact));
 }
 
 INSTANTIATE_TEST_SUITE_P(
     AstConstructionTest, AstConstructionTest,
     testing::Values(
         QueryTestData(
-            {R"(may_notifierA = query "UserA" says mayA("NotifierA"))",
-             "may_notifierA", R"("UserA")", "mayA",
-             datalog::Predicate("mayA", {R"("NotifierA")"},
-                                datalog::Sign::kPositive)}),
-        QueryTestData(
-            {R"(may_notifierB = query "UserB" says !mayB("NotifierB"))",
-             "may_notifierB", R"("UserB")", "mayB",
-             datalog::Predicate("mayB", {R"("NotifierB")"},
-                                datalog::Sign::kNegated)})));
+            {R"(may_notifierA = query "UserA" says mayA(NotifierA))",
+             "may_notifierA", Principal(R"("UserA")"),
+             Fact({}, BaseFact(datalog::Predicate("mayA", {"NotifierA"},
+                                                  datalog::Sign::kPositive)))}),
+        QueryTestData({R"(may_notifierB = query "UserB" says !mayB(NotifierB))",
+                       "may_notifierB", Principal(R"("UserB")"),
+                       Fact({}, BaseFact(datalog::Predicate(
+                                    "mayB", {"NotifierB"},
+                                    datalog::Sign ::kNegated)))})));
```

Comment:
Instead of creating a `Fact`, simply add the string representation as an expectation for predicate? e.g., 
```C++
QueryTestData(
  {R"(may_notifierA = query "UserA" says mayA(NotifierA))",
    "may_notifierA",  "UserA",  R"(mayA("NotifierA"))"},
QueryTestData(
  {R"(may_notifierB = query "UserB" says !mayB(NotifierB))",
   "may_notifierB", Principal(R"("UserB")"), R("!mayB("NotifierB")")}));
```